File: secretomeqc/pca.py
from __future__ import annotations

from dataclasses import dataclass

import numpy as np
import pandas as pd
from sklearn.decomposition import PCA
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler

from .ingest import AbundanceMatrix, log2_transform


@dataclass
class PCAResult:
    scores: pd.DataFrame
    variance: list[float]
    between_within_ratio: float
    verdict: str
    verdict_detail: str


def _lot_geometry(scores: pd.DataFrame) -> float:
    lots = scores["lot"].unique()
    centroids = scores.groupby("lot")[["pc1", "pc2"]].mean()
    within = []
    for lot in lots:
        pts = scores.loc[scores["lot"] == lot, ["pc1", "pc2"]].to_numpy()
        if len(pts) < 2:
            continue
        center = centroids.loc[lot].to_numpy()
        within.append(float(np.mean(np.linalg.norm(pts - center, axis=1))))
    mean_within = float(np.mean(within)) if within else np.nan
    if len(centroids) < 2 or not np.isfinite(mean_within) or mean_within == 0:
        return np.nan
    # Mean pairwise centroid distance / mean within-lot radius.
    coords = centroids.to_numpy()
    dists = []
    for i in range(len(coords)):
        for j in range(i + 1, len(coords)):
            dists.append(float(np.linalg.norm(coords[i] - coords[j])))
    return float(np.mean(dists) / mean_within)


def run_pca(data: AbundanceMatrix) -> PCAResult:
    log_mat = log2_transform(data.intensities)
    x = log_mat.T.to_numpy(dtype=float)
    x = SimpleImputer(strategy="median").fit_transform(x)
    x = StandardScaler().fit_transform(x)
    model = PCA(n_components=2, random_state=7)
    pcs = model.fit_transform(x)
    scores = pd.DataFrame(
        {
            "sample": log_mat.columns,
            "pc1": pcs[:, 0],
            "pc2": pcs[:, 1],
            "lot": data.sample_meta.loc[log_mat.columns, "lot"].to_numpy(),
        }
    )
    ratio = _lot_geometry(scores)
    if not np.isfinite(ratio):
        verdict, detail = "REVIEW", "Not enough replicates to score PCA lot overlap."
    elif ratio <= 2.2:
        verdict, detail = (
            "PASS",
            f"Lots overlap in PC space (between/within distance ratio {ratio:.2f} ≤ 2.2). "
            "Global abundance profiles are comparable.",
        )
    elif ratio <= 3.5:
        verdict, detail = (
            "REVIEW",
            f"Lots are partially separated in PC space (ratio {ratio:.2f}). "
            "Inspect drifting proteins before calling full comparability.",
        )
    else:
        verdict, detail = (
            "FAIL",
            f"Lots form distinct clusters (ratio {ratio:.2f} > 3.5). "
            "Global secretome profiles are not comparable.",
        )
    return PCAResult(
        scores=scores,
        variance=[float(v * 100.0) for v in model.explained_variance_ratio_],
        between_within_ratio=ratio,
        verdict=verdict,
        verdict_detail=detail,
    )

File: web/src/lib/pca.ts
import { frobeniusNormSquared, jacobiEigen, matMul, mean, median, populationStd, transpose } from "./math";
import { log2Transform } from "./ingest";
import type { AbundanceMatrix, PCAResult, Verdict } from "./types";

function imputeMedian(rows: number[][]): number[][] {
  if (!rows.length) return rows;
  const nCols = rows[0].length;
  const medians = Array.from({ length: nCols }, (_, col) => median(rows.map((row) => row[col])));
  return rows.map((row) => row.map((value, col) => (Number.isFinite(value) ? value : medians[col])));
}

function standardScale(rows: number[][]): number[][] {
  if (!rows.length) return rows;
  const nCols = rows[0].length;
  const means = Array.from({ length: nCols }, (_, col) => mean(rows.map((row) => row[col])));
  const stds = Array.from({ length: nCols }, (_, col) => {
    const s = populationStd(rows.map((row) => row[col]));
    return !Number.isFinite(s) || s === 0 ? 1 : s;
  });
  return rows.map((row) => row.map((value, col) => (value - means[col]) / stds[col]));
}

function lotGeometry(scores: { pc1: number; pc2: number; lot: string }[]): number {
  const lots = [...new Set(scores.map((row) => row.lot))];
  const centroids = new Map<string, { pc1: number; pc2: number }>();
  for (const lot of lots) {
    const pts = scores.filter((row) => row.lot === lot);
    centroids.set(lot, {
      pc1: mean(pts.map((p) => p.pc1)),
      pc2: mean(pts.map((p) => p.pc2)),
    });
  }
  const within: number[] = [];
  for (const lot of lots) {
    const pts = scores.filter((row) => row.lot === lot);
    if (pts.length < 2) continue;
    const center = centroids.get(lot)!;
    within.push(mean(pts.map((p) => Math.hypot(p.pc1 - center.pc1, p.pc2 - center.pc2))));
  }
  const meanWithin = mean(within);
  if (centroids.size < 2 || !Number.isFinite(meanWithin) || meanWithin === 0) return Number.NaN;
  const coords = [...centroids.values()];
  const dists: number[] = [];
  for (let i = 0; i < coords.length; i += 1) {
    for (let j = i + 1; j < coords.length; j += 1) {
      dists.push(Math.hypot(coords[i].pc1 - coords[j].pc1, coords[i].pc2 - coords[j].pc2));
    }
  }
  return mean(dists) / meanWithin;
}

function pcaVerdict(ratio: number): { verdict: Verdict; detail: string } {
  if (!Number.isFinite(ratio)) {
    return { verdict: "REVIEW", detail: "Not enough replicates to score PCA lot overlap." };
  }
  if (ratio <= 2.2) {
    return {
      verdict: "PASS",
      detail: `Lots overlap in PC space (between/within distance ratio ${ratio.toFixed(2)} ≤ 2.2). Global abundance profiles are comparable.`,
    };
  }
  if (ratio <= 3.5) {
    return {
      verdict: "REVIEW",
      detail: `Lots are partially separated in PC space (ratio ${ratio.toFixed(2)}). Inspect drifting proteins before calling full comparability.`,
    };
  }
  return {
    verdict: "FAIL",
    detail: `Lots form distinct clusters (ratio ${ratio.toFixed(2)} > 3.5). Global secretome profiles are not comparable.`,
  };
}

export function runPca(matrix: AbundanceMatrix): PCAResult {
  const logMat = log2Transform(matrix.intensities);
  const samplesByProtein = transpose(logMat);
  const imputed = imputeMedian(samplesByProtein);
  const scaled = standardScale(imputed);
  const gram = matMul(scaled, transpose(scaled));
  const { values, vectors } = jacobiEigen(gram);
  const order = values.map((value, idx) => ({ value, idx })).sort((a, b) => b.value - a.value);
  const totalVar = frobeniusNormSquared(scaled) || 1;
  const top = [order[0], order[1] ?? order[0]];
  const scores = matrix.samples.map((sample, row) => {
    const u1 = vectors[row][top[0].idx];
    const u2 = vectors[row][top[1].idx];
    const s1 = Math.sqrt(Math.max(top[0].value, 0));
    const s2 = Math.sqrt(Math.max(top[1].value, 0));
    return {
      sample,
      pc1: u1 * s1,
      pc2: u2 * s2,
      lot: matrix.lots[row],
    };
  });
  const ratio = lotGeometry(scores);
  const { verdict, detail } = pcaVerdict(ratio);
  return {
    scores,
    variance: [(top[0].value / totalVar) * 100, (top[1].value / totalVar) * 100],
    betweenWithinRatio: Number.isFinite(ratio) ? ratio : null,
    verdict,
    verdictDetail: detail,
  };
}
